File: examples/training/level2-reference/native_code/physics_ii.f90
subroutine physics_ii(equilibrium_in,equilibrium_out,codeparam,error_flag,error_message)

  ! ---------------------------------------
  ! PURPOSE: SIMPLE PSEUDO PHYSICS CODE 
  ! (IT READS AN EQUILIBRIUM IDS,
  ! MODIFIES ONE VARIABLE IN THE IDS, 
  ! AND WRITES THE NEW EQUILIBRIUM IDS)
  ! ---------------------------------------

  use ids_schemas, only: ids_equilibrium,ids_parameters_input,ids_is_valid
  use ids_routines, only: ids_copy
  use codeparam_input_physics

  implicit none

  type(ids_equilibrium):: equilibrium_in,equilibrium_out
  type(ids_parameters_input) :: codeparam
  type(type_codeparam_physics_data):: codeparam_data
  integer, intent(out) :: error_flag
  character(len=:), pointer, intent(out) :: error_message

  ! INITIALISATION OF ERROR FLAG
  error_flag = 0

  ! INITIAL DISPLAY
  write(*,*) ' '
  write(*,*) '======================================='
  write(*,*) 'START OF PHYSICS CODE'

  ! CHECK IF INPUT IDS IS VALID
  if (ids_is_valid(equilibrium_in%ids_properties%homogeneous_time).eq..True. &
       .and.size(equilibrium_in%time)>0) then

     call assign_codeparam(codeparam%parameters_value,codeparam_data)

     write(*,*) '------------------------------------'
     write(*,*) 'Parameters read from input xml file:'
     write(*,'(a25,i3)')   ' ntimes                = ',  &
          codeparam_data%ntimes
     write(*,'(a25,f7.3)') ' multiplication_factor = ',&
          codeparam_data%multiplication_factor
     write(*,*) '------------------------------------'

     ! COPY THE INPUT IDS IN THE OUTPUT IDS
     call ids_copy(equilibrium_in,equilibrium_out)

     ! INITIAL PLASMA MAJOR RADIUS
     write(*,'(a31,f7.3)') ' Initial plasma major radius = ', &
          equilibrium_in%time_slice(1)%boundary%geometric_axis%r

     ! MODIFY PLASMA MAJOR RADIUS
     equilibrium_out%time_slice(1)%boundary%geometric_axis%r = &
          equilibrium_out%time_slice(1)%boundary%geometric_axis%r &
          * codeparam_data%ntimes * codeparam_data%multiplication_factor

     ! MANDATORY FLAG (UNIFORM TIME HERE)
     equilibrium_out%ids_properties%homogeneous_time = 1

     ! FINAL PLASMA MAJOR RADIUS
     write(*,'(a31,f7.3)') ' Final plasma major radius   = ', &
          equilibrium_out%time_slice(1)%boundary%geometric_axis%r

  else

     ! ERROR IF THE CODE DOES NOT COMPLETE TO THE END
     error_flag = -1
     allocate(character(50):: error_message)
     error_message = 'Error in physics_ii: input IDS not valid'

  endif

  ! FINAL DISPLAY
  write(*,*) 'END OF PHYSICS CODE'
  write(*,*) '======================================='
  write(*,*) ' '

end subroutine physics_ii

File: examples/level2_cpp/native_code/physics_ii.cpp
#include <string>
#if AL_MAJOR == 5
    #include "ALClasses.h"
#elif AL_MAJOR == 4
    #include "UALClasses.h"
#else
    #warning Could not find AL_MAJOR variable. Assuming AL version = 5.x.x.
    #include "ALClasses.h"
#endif

#include "codeparam_input_physics.h"

void physics_ii_cpp(const IdsNs::IDS::equilibrium& in_equilibrium, IdsNs::IDS::equilibrium& out_equilibrium, IdsNs::codeparam_t codeparam, int& status_code, std::string& status_message)
{

    // ---------------------------------------
    // PURPOSE: SIMPLE PSEUDO PHYSICS CODE 
    // (IT READS AN EQUILIBRIUM IDS,
    // MODIFIES ONE VARIABLE IN THE IDS, 
    // AND WRITES THE NEW EQUILIBRIUM IDS)
    // ---------------------------------------
    const char* ERROR_MESSAGE = "Error in physics_ii: input IDS not valid";
    codeparam_physics_data_t codeparam_data;
    
    int idsSize = -1;
    int idsTimeMode = IDS_TIME_MODE_UNKNOWN;
    
    // INITIALISATION OF ERROR FLAG
    status_code = 0;
    
    // INITIAL DISPLAY
    printf( " \n");
    printf( "=======================================\n");
    printf( "START OF PHYSICS CODE\n");
    
    // CHECK IF INPUT IDS IS VALID
    idsSize = in_equilibrium.time.extent(0);
    idsTimeMode = in_equilibrium.ids_properties.homogeneous_time;
    if ( idsTimeMode != IDS_TIME_MODE_HOMOGENEOUS && idsSize > 0)
    {   
        // ERROR IF THE CODE DOES NOT COMPLETE TO THE END
        status_code = -1;
        status_message = ERROR_MESSAGE;
        return;
    }

    printf("XML: %s", *(codeparam.parameters));
    assign_codeparam( *(codeparam.parameters), codeparam_data);
    
    printf( "------------------------------------\n");
    printf( "Parameters read from input xml file:\n");
    printf(" ntimes                = %d\n",  codeparam_data.ntimes);
    printf(" multiplication_factor = %f\n",  codeparam_data.multiplication_factor);
    
    printf( "------------------------------------\n");
    
    // COPY THE INPUT IDS IN THE OUTPUT IDS
    out_equilibrium = in_equilibrium;
    
    // INITIAL PLASMA MAJOR RADIUS
    printf("Initial plasma major radius = %f\n", in_equilibrium.time_slice(0).boundary.geometric_axis.r);
    
    // MODIFY PLASMA MAJOR RADIUS
    out_equilibrium.time_slice.resize(1);
    double r = in_equilibrium.time_slice(0).boundary.geometric_axis.r;
    r = r * codeparam_data.ntimes * codeparam_data.multiplication_factor;
    out_equilibrium.time_slice(0).boundary.geometric_axis.r = r;
    
    // MANDATORY FLAG (UNIFORM TIME HERE)
    out_equilibrium.ids_properties.homogeneous_time = 1;
    out_equilibrium.time.resize(idsSize);
    out_equilibrium.time = in_equilibrium.time;

    // FINAL PLASMA MAJOR RADIUS
    printf(" Final plasma major radius   = %f\n",  out_equilibrium.time_slice(0).boundary.geometric_axis.r);
    
    
    // FINAL DISPLAY
    printf( "END OF PHYSICS CODE\n");
    printf( "=======================================\n");
    printf( " \n");
}
